design/design-doc-observe-only-resources.md

```diff
@@ -0,0 +1,293 @@
+two workarounds used by the community as an interim solution for this gap:
+
+- Using the provider-terraform to observe resources with the help of Terraform data sources.
+- Wrapping resources with provider-kubernetes to observe resources managed by Crossplane but as a shared object between
+multiple Compositions.
+
+In this document, we aim to introduce a solution to observe resources with Crossplane without taking ownership of them.
+This would allow users to integrate existing cloud resources with the Crossplane ecosystem without giving full ownership.
+
+## Goals
+
+- Introduce a way to observe existing resources without taking ownership of them in Crossplane.
+- Enable seamless integration of existing cloud resources with the Crossplane ecosystem.
+- Allow users to fetch data from existing resources without giving full ownership.
+
+### Non-goals
+
+- Partially managing a resource by observing a subset of fields.
+
+## Proposal
+
+To support observing resources without taking ownership, we will introduce a new spec named `managementPolicy` to the
+Managed Resources. We will also deprecate the existing `deletionPolicy` in favor of the new spec since they will be
+controlling the same behavior; that is, how should the changes on the CR affect the external cloud resource.
+
+This new policy will have the following values:
+
+- `FullControl`(Default): Crossplane will fully manage and control the external resource, including deletion when the CR
+is deleted (same as `deletionPolicy: Delete`).
+- `ObserveCreateUpdate`: Crossplane will observe and perform create and update operations on the external resource but
+will not make any deletions (same as `deletionPolicy: Orphan`).
+- `ObserveOnly`: Crossplane will only observe the external resource and will not make any changes or deletions.
+
+Please note while `ObserveCreateUpdate` may sound verbose, it accurately describes the actions that Crossplane will take
+on the external resource. This naming convention also focuses on what Crossplane will do, rather than what it won't do,
+making it more intuitive and extensible for future policy options, such as `ObserveDelete`.
+
+As indicated above, `FullControl` and `ObserveCreateUpdate` policies will behave precisely the same as the deletion
+policies we have today, including keeping the default behavior the same. We will introduce the new behavior with
+the `ObserveOnly` option, which would be pretty similar to what we have today to [import existing managed resources],
+but instead of starting to manage after import, we will not make any modifications to the external resource and only
+sync status back.
+
+```yaml
+apiVersion: ec2.aws.crossplane.io/v1beta1
+kind: VPC
+metadata:
+  annotations:
+    crossplane.io/external-name: vpc-12345678
+  name: observe-vpc
+spec:
+  managementPolicy: ObserveOnly
+  forProvider:
+    region: us-east-1
```

Comment:
I think there's a lot going for this implementation where we use the same CRs whether something is observe only or not, as opposed to introducing new CRs:

* Doing so should make migration smoother given that you update an existing CR to do so rather than replacing an existing CR with a new one.
* Existing cross-resource-references keep working.
* We avoid roughly doubling the number of CRDs we add to a system.

I will note that I don't think the observe only -> full control migration would be _dramatically_ worse if a different type was involved, and that I think we need to deprecate cross resource references in their current form regardless.

That all said, I'm struggling to get past the fact that `spec.forProvider` and `status.atProvider`, and arguably any distinction between `spec` and `status` just don't make sense when the management policy is configurable. Why make desired state (`forProvider`) distinct from observed state when desired state can be toggled to actually be observed state?

Reply:
Throwing out a possibly half-baked option that I haven't seen discussed.

Back when we designed the current managed resource schema (introducing `forProvider` etc) @muvaf proposed that `status.atProvider` in fact be a superset of `spec.forProvider`. That is:

* `spec.forProvider` include only fields that a human may configure.
* `status.atProvider` include all the fields of `spec.forProvider`, plus "output only" fields.

I believe @muvaf's rationale was that it's possible that our desired state (`forProvider`) can't actually be actuated, and that under today's schema you wouldn't have any way of knowing what the actual value was in that case. Say for example you have a database with 10GB of storage. You set `spec.forProvider.dbSize` to 20, but Crossplane can't make that change for some reason. Under the current schema you would have no way of knowing what the actual state of the database is - you'd only know Crossplane _wants_ it to have 20GB. If we had a corresponding `status.atProvider.dbSize` field you could see what was different.

At the time I argued against duplicating spec fields under status because it was not an 'idiomatic' Kubernetes CR pattern. I'm not sure that logic holds up anymore though - there's no precedence that I'm aware of for functionality like observe only resources either, but we still need to build it.

The reason I share all of this context is that I wonder whether duplicating spec fields under `status.atProvider` would be a good observe only UX:

* _Everything_ we can observe about a resource appears under `status.atProvider` regardless of the management policy of the resource. We update any cross resource references that currently read from `spec.forProvider` to instead read from the equivalent `status.atProvider` fields.
* If you want to specify desired state, include it in `spec.forProvider`. If you don't, don't include it. You could almost omit the `managementPolicy` under this approach, but I think it would still be useful to control the schema (requiredness) of `spec.forProvider` fields.

Reply:
I like having `status.atProvider` to include the full state. 
I cannot think of any immediate drawbacks of this approach, and makes sense considering the advantages you already have listed.

Reply:
Hmm, one caveat is some possible friction with migration. 
Currently, the scenario is, to start with `managementPolicy: ObserveOnly`, and once you feel good, simply set `managementPolicy: Full` and expect XP automagically start managing.
If we have all observations in `status.atProvider`, the required ones need to be moved up to `spec.forProvider` in addition to changing the `managementPolicy` to `Full`.

Reply:
> Would we need a managementPolicy at all in this scenario?

`managementPolicy` would at least be needed to decide whether a field is required (with CEL), e.g. if the `managementPolicy: Full` you need to provide `spec.size`. 

I also have some concerns about moving from resource ownership to a field ownership model but I need to spend some time to materialize them with some examples.

Reply:
The primary concern I have with moving from _resource ownership_ to a _field ownership_ model is that it would be a bold move since it is changing something fundamental. One example consequence of such a change is, late-initialization logic would need to be revisited for all resources, and we would probably late-init to `status.atProvider` and do some merging with `spec.forProvider` before the update calls.

I rather prefer to defer it to a dedicated design, e.g. field level management, if it solves some real-world problems. So far, the only use case that could be relevant is [this issue](https://github.com/crossplane/crossplane/issues/3516), which could have more straightforward solutions like the one proposed in the issue or as you proposed [here](https://github.com/crossplane/crossplane/pull/3531#discussion_r1072945963). 

Even if we go with that route, I believe having a `managementPolicy` would be helpful since:
- Being explicit about marking a resource as Read-Only would make folks more comfortable while adopting existing resources.
- We can use it to decide whether a field is required (with CEL).

So, I would suggest continuing with the proposed design and considering moving to a more granular solution if we see such a need.

Reply:
I agree this is a big change, but if we think it's the better way to do things longer term I would rather start moving toward that than introduce an "intermediate" version of the feature . If we add a management policy that puts `spec.forProvider` into "observe only mode" that's behaviour (and potentially a few field/knob) that we'd have to deprecate and remove if we moved to including everything in `status.atProvider` eventually.

Reply:
Ok, after putting in some more thoughts, I think I can now articulate my concerns with an example.

The API without `managementPolicy` lacks information between the MR and the cloud resource despite it having that for individual fields. As an example, let's think about the reconciler logic operating on the following resource:

```yaml
apiVersion: eks.aws.crossplane.io/v1beta1
kind: Cluster
metadata:
  name: some-eks-cluster
  annotations:
    crossplane.io/external-name: some-eks-cluster
spec:
  forProvider:
    region: us-west-1
```

Our reconciler will read the MR from k8s API and try to get the EKS cluster named `some-eks-cluster` in `us-west-1` from AWS API. If it exists, it will check if all the spec matches what is provided by the user under `spec.forProvider`, and otherwise do an update call to AWS to get the desired state synced.

However, **if the cluster does NOT exist, it is not clear what to do**:
- If the user wants an EKS cluster with all the default parameters, **we need to create the cluster**.
- If the user's intent is to observe an existing EKS cluster, **we need to fail with a proper error**.

This is because we no longer know the relationship between MR and the resource itself, even though we can have it at the field level. 

The bottom line is we need to have the `managementPolicy` field to infer that relationship.

Reply:
> - If you're creating a new resource you must include the minimum required set of fields in spec.forProvider.
> - If you're observing/importing an existing resource, you can omit all fields from spec.forProvider.

Re-reading your thoughts above and I believe you have the assumption that all resources will always have at least one required field, which could indeed be used to distinguish create vs observe. However, this does not hold for most of the resources, and even if it were, it would be quite an implicit distinction.

Reply:
@turkenh You're right, I don't think we can assume an empty `spec.forProvider` means someone wants an observe-only resource. So we do need `managementPolicy` (or something like it).

I still feel quite strongly that it shouldn't be possible to put `spec.forProvider` into an "observe only mode" though. I would like `spec.forProvider` to _only_ ever be used for desired state, with all observed state in `status.atProvider`. If we keep the management policy, are there any outstanding issues with this approach?

I know you pointed out in https://github.com/crossplane/crossplane/pull/3531#discussion_r1068122286 that toggling from observe-only to not would require editing the MR move stuff up from `status.atProvider` to `spec.forProvider`. Presumably if we went the other way (Full -> OO) we'd have to remove stuff from `spec.atProvider` too. Do you consider this a show-stopper?

Reply:
@negz as we discussed offline, I've updated the design to have a full observed state under `status.atProvider`  https://github.com/crossplane/crossplane/pull/3531/commits/be9d560dda3aec3912d61c58fc5efbc8e93aa330